### DISCUSSION TITLE
Workaround for specifying global metadata if layout.tsx is a client component

### DISCUSSION THREAD

**Discussion by sebpowell** [vercel/next.js#47547] [Help] [answered]
### Summary

I've recently upgraded my project to Next 13 and running into the following problem. 

My `layout.tsx` file, located in the root `app/` folder, loads in two components that use `useEffect` (a navigation bar and a footer). As a result, I've had to mark the `layout.tsx` file as a client component. 

So my file looks something like this:

```TSX
"use client";
import "../app/global.scss";
import { Footer } from "@/components/organisms/Footer";
import { Navigation } from "@/components/organisms/Navigation";

export default function RootLayout({ children }: { children: ReactNode }) {

  return (
    <html lang="en" style={{ scrollBehavior: "smooth" }}>
      <body>
        <Providers>
          <Navigation />
          {children}
          <Footer />
        </Providers>
        <Scripts />
      </body>
    </html>
  );
}
```

If I try specifying a default metadata object, as documented here https://beta.nextjs.org/docs/api-reference/metadata, it doesn't work – apparently, this can only be used in server components. 

```TSX
"use client";
import "../app/global.scss";
import { Footer } from "@/components/organisms/Footer";
import { Navigation } from "@/components/organisms/Navigation";

// This doesn't work – "The Next.js 'metadata' API is not allowed in a client component."
export const metadata = {
  title: "Site title",
};

export default function RootLayout({ children }: { children: ReactNode }) {

  return (
    <html lang="en" style={{ scrollBehavior: "smooth" }}>
      <body>
        <Providers>
          <Navigation />
          {children}
          <Footer />
        </Providers>
        <Scripts />
      </body>
    </html>
  );
}
```

So I'm wondering if there's a workaround, or if I just need to set meta attributes in every file individually (I liked the idea of being able to use the title templates feature, but might not be possible). 

Any ideas gratefully received. 

### Additional information

Using Next 13.2.1

**Comment by tmitchel2** [comment]
I've literally just had this same issue... I would be interested to know the best practise for this.

**Comment by icyJoseph** [comment, marked as answer]
Does this work?

```tsx
"use client";
// src/components/foo-page-wrapper-something.tsx

import type { ReactNode } from "react";
import { Footer } from "@/components/organisms/Footer";
import { Navigation } from "@/components/organisms/Navigation";

export const AppWrapper = ({ children }: { children: ReactNode }) => {
  return (
    <Providers>
      <Navigation />
      {children}
      <Footer />
    </Providers>
  );
};
```

And then:

```tsx
// app/layout.tsx

import type { ReactNode } from "react";
import { AppWrapper } from "@/components/foo-app-wrapper-something";

// define meta here
export const metadata = {
  title: "Create Next App",
  description: "Generated by create next app",
};

export default function RootLayout({ children }: { children: ReactNode }) {
  return (
    <html lang="en" style={{ scrollBehavior: "smooth" }}>
      <body>
        <AppWrapper>{children}</AppWrapper>
        <Scripts />
      </body>
    </html>
  );
}
```

**Reply by sebpowell** [reply]
@icyJoseph – thanks for the suggestions Joseph – sadly not, at least, not in my case. It seems that if any child components use any effects, Next expects the layout component to be marked as a client component. I tried doing the above, and getting the following error:

```
You're importing a component that needs useState. It only works in a Client Component but none of its parents are marked with "use client", so they're Server Components by default.

  8 | import { NavigationHamburger } from "./NavigationHamburger";
  9 | import { useDisclosure } from "hooks/use-disclosure";
 10 | import { useWindowScroll } from "react-use";
 11 | import { useEffect, useRef, useState } from "react";
    :                             ^^^^^^^^
 12 | import { NavigationDesktopPrimaryLinks } from "./NavigationDesktop/NavigationDesktopPrimaryLinks";
 13 | import { NavigationContainer } from "./NavigationContainer";
 14 | import { usePathname } from "next/navigation";
    `----
    
    ```

**Reply by icyJoseph** [reply]
Mm I tried this one and it worked.

Could you share more code? 

In some cases layout must be client but not always. As far as I remember you can always create your modules in such a way that they encapsulate the client entry from their parent.

**Reply by masterbater** [reply]
> @icyJoseph – thanks for the suggestions Joseph – sadly not, at least, not in my case. It seems that if any child components use any effects, Next expects the layout component to be marked as a client component. I tried doing the above, and getting the following error:
> 
> ```
> You're importing a component that needs useState. It only works in a Client Component but none of its parents are marked with "use client", so they're Server Components by default.
> 
>   8 | import { NavigationHamburger } from "./NavigationHamburger";
>   9 | import { useDisclosure } from "hooks/use-disclosure";
>  10 | import { useWindowScroll } from "react-use";
>  11 | import { useEffect, useRef, useState } from "react";
>     :                             ^^^^^^^^
>  12 | import { NavigationDesktopPrimaryLinks } from "./NavigationDesktop/NavigationDesktopPrimaryLinks";
>  13 | import { NavigationContainer } from "./NavigationContainer";
>  14 | import { usePathname } from "next/navigation";
>     `----
>     
>     ```

Probably you haven't removed the imports.

**Reply by sebpowell** [reply]
@icyJoseph – I moved things around, and got it to work, thank you so much! (and you @masterbater). @tmitchel2 – see above suggestion, this sorted it.

**Reply by icyJoseph** [reply]
Nice! Please don't forget to mark as answered 😅🎉

**Reply by tt-alex** [reply]
does it mean this is still a work around, just to put title tag to work at "use client" pages... too much work for insignificant strings....
anyone know does nextjs team are working on a fix for this or will gonna use workaround inpermanent? hope they wont make their "stable" version as suck as angular2 =-/
